H2 not working with file based databases

### Task List

- [x] Steps to reproduce provided
- [x] Stacktrace (if present) provided
- [x] Example that reproduces the problem uploaded to Github
- [x] Full description of the issue provided (see below)

### Steps to Reproduce

1. build native image of the example application
2. run the native image

### Expected Behaviour

Application creates file based database

### Actual Behaviour

Application crashes at startup
<details>
<summary>stacktrace</summary>

```
←[36m14:35:53.441←[0;39m ←[1;30m[main]←[0;39m ←[34mINFO ←[0;39m ←[35mcom.zaxxer.hikari.HikariDataSource←[0;39m - HikariPool-1 - Starting...

org.h2.jdbc.JdbcSQLNonTransientException: General error: "java.lang.NullPointerException" [50000-199]
        at org.h2.engine.SessionRemote.connectEmbeddedOrServer(SessionRemote.java:338)
        at org.h2.jdbc.JdbcConnection.<init>(JdbcConnection.java:148)
        at org.h2.Driver.connect(Driver.java:69)
        at com.zaxxer.hikari.util.DriverDataSource.getConnection(DriverDataSource.java:138)
        at com.zaxxer.hikari.pool.PoolBase.newConnection(PoolBase.java:358)
        at com.zaxxer.hikari.pool.PoolBase.newPoolEntry(PoolBase.java:206)
        at com.zaxxer.hikari.pool.HikariPool.createPoolEntry(HikariPool.java:477)
        at com.zaxxer.hikari.pool.HikariPool.checkFailFast(HikariPool.java:560)
        at com.zaxxer.hikari.pool.HikariPool.<init>(HikariPool.java:115)
        at com.zaxxer.hikari.HikariDataSource.<init>(HikariDataSource.java:81)
        at io.micronaut.configuration.jdbc.hikari.HikariUrlDataSource.<init>(HikariUrlDataSource.java:35)
        at io.micronaut.configuration.jdbc.hikari.DatasourceFactory.dataSource(DatasourceFactory.java:66)
        at io.micronaut.configuration.jdbc.hikari.$DatasourceFactory$DataSource0Definition.build(Unknown Source)
        at io.micronaut.context.BeanDefinitionDelegate.build(BeanDefinitionDelegate.java:143)
        at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1888)
        at io.micronaut.context.DefaultBeanContext.createAndRegisterSingletonInternal(DefaultBeanContext.java:2630)
        at io.micronaut.context.DefaultBeanContext.createAndRegisterSingleton(DefaultBeanContext.java:2616)
        at io.micronaut.context.DefaultBeanContext.loadContextScopeBean(DefaultBeanContext.java:2166)
        at io.micronaut.context.DefaultBeanContext.initializeContext(DefaultBeanContext.java:1531)
        at io.micronaut.context.DefaultApplicationContext.initializeContext(DefaultApplicationContext.java:220)
        at io.micronaut.context.DefaultBeanContext.readAllBeanDefinitionClasses(DefaultBeanContext.java:2763)
        at io.micronaut.context.DefaultBeanContext.start(DefaultBeanContext.java:227)
        at io.micronaut.context.DefaultApplicationContext.start(DefaultApplicationContext.java:166)
        at io.micronaut.runtime.Micronaut.start(Micronaut.java:64)
        at io.micronaut.runtime.Micronaut.run(Micronaut.java:294)
        at io.micronaut.runtime.Micronaut.run(Micronaut.java:280)
        at com.example.Application.main(Application.java:8)
Caused by: java.lang.NullPointerException: null
        at org.h2.store.fs.FilePath.get(FilePath.java:54)
        at org.h2.store.fs.FileUtils.isAbsolute(FileUtils.java:102)
        at org.h2.engine.ConnectionInfo.getName(ConnectionInfo.java:383)
        at org.h2.engine.Engine.openSession(Engine.java:53)
        at org.h2.engine.Engine.openSession(Engine.java:201)
        at org.h2.engine.Engine.createSessionAndValidate(Engine.java:178)
        at org.h2.engine.Engine.createSession(Engine.java:161)
        at org.h2.engine.Engine.createSession(Engine.java:31)
        at org.h2.engine.SessionRemote.connectEmbeddedOrServer(SessionRemote.java:336)
        ... 27 common frames omitted
←[36m14:35:54.447←[0;39m ←[1;30m[main]←[0;39m ←[1;31mERROR←[0;39m ←[35mio.micronaut.runtime.Micronaut←[0;39m - Error starting Micronaut server: Bean definition [javax.sql.DataSource] could not be loaded: Error instantiating bean of type [javax.sql.DataSource]: Failed to initialize pool: General error: "java.lang.NullPointerException" [50000-199]
io.micronaut.context.exceptions.BeanInstantiationException: Bean definition [javax.sql.DataSource] could not be loaded: Error instantiating bean of type [javax.sql.DataSource]: Failed to initialize pool: General error: "java.lang.NullPointerException" [50000-199]
        at io.micronaut.context.DefaultBeanContext.initializeContext(DefaultBeanContext.java:1533)
        at io.micronaut.context.DefaultApplicationContext.initializeContext(DefaultApplicationContext.java:220)
        at io.micronaut.context.DefaultBeanContext.readAllBeanDefinitionClasses(DefaultBeanContext.java:2763)
        at io.micronaut.context.DefaultBeanContext.start(DefaultBeanContext.java:227)
        at io.micronaut.context.DefaultApplicationContext.start(DefaultApplicationContext.java:166)
        at io.micronaut.runtime.Micronaut.start(Micronaut.java:64)
        at io.micronaut.runtime.Micronaut.run(Micronaut.java:294)
        at io.micronaut.runtime.Micronaut.run(Micronaut.java:280)
        at com.example.Application.main(Application.java:8)
Caused by: io.micronaut.context.exceptions.BeanInstantiationException: Error instantiating bean of type [javax.sql.DataSource]: Failed to initialize pool: General error: "java.lang.NullPointerException" [50000-199]
        at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1917)
        at io.micronaut.context.DefaultBeanContext.createAndRegisterSingletonInternal(DefaultBeanContext.java:2630)
        at io.micronaut.context.DefaultBeanContext.createAndRegisterSingleton(DefaultBeanContext.java:2616)
        at io.micronaut.context.DefaultBeanContext.loadContextScopeBean(DefaultBeanContext.java:2166)
        at io.micronaut.context.DefaultBeanContext.initializeContext(DefaultBeanContext.java:1531)
        ... 8 common frames omitted
Caused by: com.zaxxer.hikari.pool.HikariPool$PoolInitializationException: Failed to initialize pool: General error: "java.lang.NullPointerException" [50000-199]
        at com.zaxxer.hikari.pool.HikariPool.throwPoolInitializationException(HikariPool.java:595)
        at com.zaxxer.hikari.pool.HikariPool.checkFailFast(HikariPool.java:581)
        at com.zaxxer.hikari.pool.HikariPool.<init>(HikariPool.java:115)
        at com.zaxxer.hikari.HikariDataSource.<init>(HikariDataSource.java:81)
        at io.micronaut.configuration.jdbc.hikari.HikariUrlDataSource.<init>(HikariUrlDataSource.java:35)
        at io.micronaut.configuration.jdbc.hikari.DatasourceFactory.dataSource(DatasourceFactory.java:66)
        at io.micronaut.configuration.jdbc.hikari.$DatasourceFactory$DataSource0Definition.build(Unknown Source)
        at io.micronaut.context.BeanDefinitionDelegate.build(BeanDefinitionDelegate.java:143)
        at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1888)
        ... 12 common frames omitted
Caused by: org.h2.jdbc.JdbcSQLNonTransientException: General error: "java.lang.NullPointerException" [50000-199]
        at org.h2.engine.SessionRemote.connectEmbeddedOrServer(SessionRemote.java:338)
        at org.h2.jdbc.JdbcConnection.<init>(JdbcConnection.java:169)
        at org.h2.jdbc.JdbcConnection.<init>(JdbcConnection.java:148)
        at org.h2.Driver.connect(Driver.java:69)
        at com.zaxxer.hikari.util.DriverDataSource.getConnection(DriverDataSource.java:138)
        at com.zaxxer.hikari.pool.PoolBase.newConnection(PoolBase.java:358)
        at com.zaxxer.hikari.pool.PoolBase.newPoolEntry(PoolBase.java:206)
        at com.zaxxer.hikari.pool.HikariPool.createPoolEntry(HikariPool.java:477)
        at com.zaxxer.hikari.pool.HikariPool.checkFailFast(HikariPool.java:560)
        ... 19 common frames omitted
Caused by: java.lang.NullPointerException: null
        at org.h2.store.fs.FilePath.get(FilePath.java:54)
        at org.h2.store.fs.FileUtils.isAbsolute(FileUtils.java:102)
        at org.h2.engine.ConnectionInfo.getName(ConnectionInfo.java:383)
        at org.h2.engine.Engine.openSession(Engine.java:53)
        at org.h2.engine.Engine.openSession(Engine.java:201)
        at org.h2.engine.Engine.createSessionAndValidate(Engine.java:178)
        at org.h2.engine.Engine.createSession(Engine.java:161)
        at org.h2.engine.Engine.createSession(Engine.java:31)
        at org.h2.engine.SessionRemote.connectEmbeddedOrServer(SessionRemote.java:336)
        ... 27 common frames omitted
```
</details>

### Environment Information

- **Operating System**: Windows 10
- **Micronaut Version:** 2.0.0.RC2
- **JDK Version:** GraalVM 20.1.0 java 11

### Example Application

https://github.com/kristofdho/hikari-gvm

The application is generated with https://micronaut.io/launch/ and following settings:
![image](https://user-images.githubusercontent.com/43339887/85720232-88b21580-b6f0-11ea-882b-985fa8b1d29c.png)

The only change is the datasource url from `mem` to `file`:
`datasources.default.url: jdbc:h2:file:${java.io.tmpdir}/demodb;MVCC=TRUE;LOCK_TIMEOUT=10000;DB_CLOSE_ON_EXIT=FALSE`
instead of the default
`datasources.default.url: jdbc:h2:mem:devDb;MVCC=TRUE;LOCK_TIMEOUT=10000;DB_CLOSE_ON_EXIT=FALSE`

### My 2 cents
caused by the `org.h2.store.fs.FilePath` class not being able to resolve its subclasses at runtime.

<details>
<summary>culprit snippet FilePath#registerDefaultProviders</summary>

```java
    private static void registerDefaultProviders() {
        if (providers == null || defaultProvider == null) {
            ConcurrentHashMap<String, FilePath> map = new ConcurrentHashMap<>();
            for (String c : new String[] {
                    "org.h2.store.fs.FilePathDisk",
                    "org.h2.store.fs.FilePathMem",
                    "org.h2.store.fs.FilePathMemLZF",
                    "org.h2.store.fs.FilePathNioMem",
                    "org.h2.store.fs.FilePathNioMemLZF",
                    "org.h2.store.fs.FilePathSplit",
                    "org.h2.store.fs.FilePathNio",
                    "org.h2.store.fs.FilePathNioMapped",
                    "org.h2.store.fs.FilePathAsync",
                    "org.h2.store.fs.FilePathZip",
                    "org.h2.store.fs.FilePathRetryOnInterrupt"
            }) {
                try {
                    FilePath p = (FilePath) Class.forName(c).getDeclaredConstructor().newInstance();
                    map.put(p.getScheme(), p);
                    if (defaultProvider == null) {
                        defaultProvider = p;
                    }
                } catch (Exception e) {
                    // ignore - the files may be excluded in purpose
                }
            }
            providers = map;
        }
    }
```
</details>

I have added those classes for runtime reflective instantiation myself, but some buildtime <-> runtime initialization issues arrise during the image build. I am looking further into this and will update the ticket when i know more.

--- a/jdbc/src/main/java/io/micronaut/jdbc/nativeimage/JdbcFeature.java
+++ b/jdbc/src/main/java/io/micronaut/jdbc/nativeimage/JdbcFeature.java
@@ -76,6 +76,20 @@ final class JdbcFeature implements Feature {
             RuntimeReflection.register(h2Driver);
             RuntimeClassInitialization.initializeAtBuildTime(h2Driver);
 
+            // required for file-based H2 databases
+            registerForRuntimeReflectiveInstantiation(access, "org.h2.store.fs.FilePathDisk");
+            registerForRuntimeReflectiveInstantiation(access, "org.h2.store.fs.FilePathMem");
+            registerForRuntimeReflectiveInstantiation(access, "org.h2.store.fs.FilePathMemLZF");
+            registerForRuntimeReflectiveInstantiation(access, "org.h2.store.fs.FilePathNioMem");
+            registerForRuntimeReflectiveInstantiation(access, "org.h2.store.fs.FilePathNioMemLZF");
+            registerForRuntimeReflectiveInstantiation(access, "org.h2.store.fs.FilePathSplit");
+            registerForRuntimeReflectiveInstantiation(access, "org.h2.store.fs.FilePathNio");
+            registerForRuntimeReflectiveInstantiation(access, "org.h2.store.fs.FilePathNioMapped");
+            registerForRuntimeReflectiveInstantiation(access, "org.h2.store.fs.FilePathAsync");
+            registerForRuntimeReflectiveInstantiation(access, "org.h2.store.fs.FilePathZip");
+            registerForRuntimeReflectiveInstantiation(access, "org.h2.store.fs.FilePathRetryOnInterrupt");
+            initializeAtRuntime(access, "sun.nio.ch.WindowsAsynchronousFileChannelImpl$DefaultIocpHolder");
+
             ResourcesRegistry resourcesRegistry = getResourceRegistry();
             if (resourcesRegistry != null) {
                 resourcesRegistry.addResources("META-INF/services/java.sql.Driver");
@@ -229,6 +243,14 @@ final class JdbcFeature implements Feature {
             for (Constructor constructor : t.getConstructors()) {
                 RuntimeReflection.register(constructor);
             }
+        }
+    }
+
+    private void registerForRuntimeReflectiveInstantiation(BeforeAnalysisAccess access, String n) {
+        Class<?> t = access.findClassByName(n);
+        if (t != null) {
+            RuntimeReflection.register(t);
+            RuntimeReflection.registerForReflectiveInstantiation(t);
         }
     }
 

